x/auth/vesting/cmd/vestcalc/vestcalc.go

```diff
@@ -0,0 +1,532 @@
+		runningTot = nextTot
+	}
+	// Integrity check
+	sum := int64(0)
+	for _, x := range a {
+		sum = sum + x
+	}
+	if sum != total {
+		return nil, fmt.Errorf("failed integrity check: divisions sum to %d, should be %d", sum, total)
+	}
+	return a, nil
+}
+
+// vestingTimes generates timestamps for successive months after startTime.
+// The monthly events occur at the given time of day. If the month is not
+// long enough for the desired date, the last day of the month is used.
+func vestingTimes(startTime time.Time, months int32, timeOfDay time.Time) ([]time.Time, error) {
+	if months < 1 {
+		return nil, fmt.Errorf("must have at least one vesting period")
+	}
+	location := startTime.Location()
+	hour := timeOfDay.Hour()
+	minute := timeOfDay.Minute()
+	second := timeOfDay.Second()
+	times := make([]time.Time, months)
+	for i := 1; i <= int(months); i++ {
+		tm := startTime.AddDate(0, int(i), 0)
+		if tm.Day() != startTime.Day() {
+			// The starting day-of-month cannot fit in this month,
+			// and we've wrapped to the next month. Back up to the
+			// end of the previous month.
+			tm = tm.AddDate(0, 0, -tm.Day())
+		}
+		times[i-1] = time.Date(tm.Year(), tm.Month(), tm.Day(), hour, minute, second, 0, location)
+	}
+	// Integrity check: dates must be sequential and 26-33 days apart.
+	lastTime := startTime
+	for _, tm := range times {
+		duration := tm.Sub(lastTime)
+		if duration < 26*24*time.Hour {
+			return nil, fmt.Errorf("vesting dates too close: %v and %v", lastTime, tm)
+		}
+		if duration > 33*24*time.Hour {
+			return nil, fmt.Errorf("vesting dates too distant: %v and %v", lastTime, tm)
+		}
+		lastTime = tm
+	}
+	return times, nil
+}
+
+// period represents a single period in the vesting schdule.
+// The length in seconds is the time to the vesting event
+// from the previous event in the file, or the start time.
+type period struct {
+	Coins  string `json:"coins"`
+	Length int64  `json:"length_seconds"`
+}
+
+// encodeCoins encodes the given amount and denomination in coin format.
+func encodeCoins(amount int64, denom string) string {
+	return fmt.Sprint(amount) + denom
+}
+
+// parseCoins decodes the coin format into an amount and denomination
+func parseCoins(coins string) (int64, string) {
+	var amount int64
+	var denom string
+	fmt.Sscanf(coins, "%d%s", &amount, &denom)
+	return amount, denom
+}
+
+// marshalPeriods gives the JSON encoding.
+func marshalPeriods(periods []period) ([]byte, error) {
+	return json.MarshalIndent(periods, "", "  ")
+}
+
+// unmarshalPeriods parses an array of periods in JSON.
+func unmarshalPeriods(bz []byte) ([]period, error) {
+	periods := []period{}
+	err := json.Unmarshal(bz, &periods)
+	if err != nil {
+		return nil, err
+	}
+	return periods, nil
+}
+
+// event represents a single vesting event with an absolute time.
+// The denomination must be understood by context.
+type event struct {
+	Time   time.Time
+	Amount int64 // TODO support 255-bit coin amounts
```

Comment:
Glad you mentioned this.  I'd specify in the TODO list: "replace int64 with sdk.Int".

Reply:
Done.